plugins/modules/cloud/docker/docker_stack_info.py

```diff
@@ -0,0 +1,122 @@
+    'supported_by': 'community',
+    'metadata_version': '1.1'
+}
+
+DOCUMENTATION = '''
+---
+module: docker_stack_info
+author: "Jose Angel Munoz (@imjoseangel)"
+short_description: docker stack module info
+description:
+  - Manage docker stacks using the 'docker stack' command
+    on the target node (see examples).
+version_added: "2.11"
+options:
+  name:
+    description:
+      - Stack name
+    type: str
+  action:
+    description:
+      - If set to C(tasks) displays the list of stacks using the C(docker stack ls) command.
+        If not set, the module use "list" by default.
+    type: str
+    choices: ["list", "tasks"]
+    default: list
+'''
+
+RETURN = '''
+msg:
+    description: |
+        dictionary containing the list of stacks or tasks associated
+        to a stack name.
+    sample: >
+        "msg": [{"name":"grafana","namespace":"default","orchestrator":"Kubernetes","services":"2"}]
+    returned: always
+    type: list
+'''
+
+EXAMPLES = '''
+  - name: Shows stack info
+    docker_stack_info:
+'''
+
+import ast
+from ansible.module_utils.basic import AnsibleModule
+from boto.dynamodb import item
+
+
+def docker_stack_list(module):
+    docker_bin = module.get_bin_path('docker', required=True)
+    rc, out, err = module.run_command([
+        docker_bin, "stack", "ls", "--format",
+        '{ "name": "{{.Name}}", "services": "{{.Services}}", "orchestrator": "{{.Orchestrator}}", "namespace": "{{.Namespace}}" }'
+    ])
+
+    return rc, out.strip(), err.strip()
+
+
+def docker_stack_tasks(module, name):
+    docker_bin = module.get_bin_path('docker', required=True)
+    rc, out, err = module.run_command([
+        docker_bin, "stack", "ps", name, "--format",
+        '{ "id": "{{.ID}}", "name": "{{.Name}}", "image": "{{.Image}}", "node": "{{.Node}}", "desired state": "{{.DesiredState}}", "current state": "{{.CurrentState}}", "error": "{{.Error}}", "ports": "{{.Ports}}" }'
+    ])
+    return rc, out.strip(), err.strip()
+
+
+def main():
+    module = AnsibleModule(argument_spec={
+        'name':
+        dict(type='str', required=False),
+        'action':
+        dict(type='str', choices=['list', 'tasks'], default="list"),
+    },
+                           supports_check_mode=False)
+
+    results = []
+    name = module.params['name']
+    action = module.params['action']
+
+    if action == 'tasks':
+        if not name:
+            module.fail_json(msg=("action requires a stack name"))
+
+        rc, out, err = docker_stack_tasks(module, name)
+
+    elif action == "list":
+        rc, out, err = docker_stack_list(module)
+
+    if rc != 0:
+        module.fail_json(msg="'docker stack down' command failed",
+                         rc=rc,
+                         out=out,
+                         err=err,
+                         stdout=out,
+                         stderr=err)
+    else:
+
+        if out:
+            results = list(ast.literal_eval(item) for item in out.split("\n"))
```

Comment:
I'm not sure whether I like this. What exactly is the tool returning? JSON? YAML? Python code?

Reply:
The tools is returning strings like:

```
ID                  NAME                                IMAGE                         NODE                DESIRED STATE       CURRENT STATE            ERROR               PORTS
ef45cb00-3c3        grafana_grafana-0                   grafana/grafana               docker-desktop      Running             Running 22 minutes ago                       *:0->3000/tcp
0809163b-065        grafana_graphite-5b5b99dd5f-td4pm   graphiteapp/graphite-statsd   docker-desktop      Running             Running 22 minutes ago                       *:0->80/tcp,*:0->2003/tcp,*:0->2004/tcp
```

What we are doing here is converting to JSON to be able to play with the data.

Reply:
Well, with a multiline result in the stack we will get the following string:

```
{"CurrentState":"Running 21 hours ago","DesiredState":"Running","Error":"","ID":"f5710960-47a","Image":"grafana/grafana","Name":"grafana_grafana-0","Node":"docker-desktop","Ports":"*:0-\u003e3000/tcp"}
{"CurrentState":"Running 21 hours ago","DesiredState":"Running","Error":"","ID":"568ee46c-c0b","Image":"graphiteapp/graphite-statsd","Name":"grafana_graphite-5b5b99dd5f-ksf9r","Node":"docker-desktop","Ports":"*:0-\u003e80/tcp,*:0-\u003e2003/tcp,*:0-\u003e2004/tcp"}
```

The conditional

```python
        if out:
            ret = list(
                json.loads(ast.literal_eval(outitem))
                for outitem in out.split("\n"))
```

Takes every string and converts to json adding it to a list that can be managed later in the ret variable.

Reply:
Let's resolve and discuss. Tests and output are now adapted